File: talleres/taller3/Main.java
import java.util.Scanner;

public class Main {

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int casos = scanner.nextInt();
        for(int i = 0; i < casos; i++){
            int t = scanner.nextInt(); 
            int h = scanner.nextInt();
            int f = scanner.nextInt();
            int[][] arboles = new int[h][t];
            // Armo una matriz en donde cada columna representa un arbol y cada fila una altura. En esta matriz guardo cuantas bellotas hay en cada altura en cada arbol.
            for(int j = 0; j < t; j++){
                int cant_bellotas = scanner.nextInt();
                for(int k = 0; k < cant_bellotas; k++){
                    arboles[scanner.nextInt() - 1][j] += 1;
                }
            }
            // Llamo a la funcion que calcula cual es la mayor cantidad de bellotas que se pueden recolectar en una bajada.
            int res = mjBajada(arboles, h, t, f);
            System.out.println(res);

        }   
        scanner.nextInt();
        scanner.close();
    }

    public static int mjBajada(int[][] arboles, int h, int t, int f){
        // Creo una lista de la misma longitud que las alturas de los arboles. En esta lista guardo en la posicion i la mayor cantidad de bellotas que se pueden recolectar bajando 
        // desde esa altura.
        int[] max_por_altura = new int[h];
        
        // Recorro para cada altura cada arbol y calculo cual es el maximo de bellotas que se pueden recolectar desde esa posicion. Luego guardo esta informacion en la matriz de los arboles
        // para no tener que calcularla nuevamente si la vuelvo a necesitar.
        for(int i = 0; i < h; i++){
            for(int j = 0; j < t; j++){
                if(i - f >= 0){
                    arboles[i][j] = Math.max(arboles[i - 1][j], max_por_altura[i -f]) + arboles[i][j];
                }
                else if(i != 0){
                    arboles[i][j] = arboles[i - 1][j] + arboles[i][j];
                }
                if(arboles[i][j] >= max_por_altura[i]){
                    max_por_altura[i] = arboles[i][j];
                }
            }
        }
        int res = max_por_altura[h-1];
        return res;
    }
}

File: talleres/taller3/main.py
def mejor_bajada(arboles, j, t, f, M):
    def maxEnPos(arboles, altura, arbol, h, t, f):
        if(altura == h - 1):
            M[altura][arbol] = 0
        if(M[altura][arbol] != -1):
            return M[altura][arbol]

        
        max_actual = 0
        for i in range(t):
            if(i != arbol and altura + f < h):
                if(M[altura + f][i] == -1):
                    M[altura + f][i] = maxEnPos(arboles, altura + f, i, h, t, f);
                
                if(M[altura+f][i] > max_actual):
                    max_actual = M[altura+f][i]

                
            elif(i == arbol):
                if(M[altura + 1][i] == -1):
                    M[altura + 1][i] = maxEnPos(arboles, altura + 1, i, h, t, f);
                 
                if(M[altura+1][i] > max_actual):
                    max_actual = M[altura+1][i]
                
        M[altura][arbol] = arboles[altura][arbol] + max_actual;
        return M[altura][arbol];




    for i in range(t):
        maxEnPos(arboles, 0, i, h, t, f)
    
arboles = [
    [1,0,0],
    [0,2,1],
    [0,1,0],
    [0,1,0],
    [0,0,1],
    [0,1,1],
    [1,0,1],
    [0,1,1],
    [0,0,0],
    [1,0,0]
]

h = 10
t = 3
f = 2
M  = [[-1] * t for _ in range(h)]
res = mejor_bajada(arboles, h, t, f, M);
for i in range(len(M)):
    print("columna: " + str(i) + " ")

    for j in range(len(M[i])):
        print(str(M[i][j]) + " ")
            

        
print(res);
        
    

